File: 2615-sum-of-distances/2615-sum-of-distances.py
class Solution:
    def distance(self, nums: List[int]) -> List[int]:
        res = []
        idxPre = {}
        idxSuf = {}
        pre = [0 for _ in range(len(nums))]
        suf = [0 for _ in range(len(nums))]
        for i in range(len(nums)):
            if nums[i] in idxPre:
                pre[i] = pre[idxPre[nums[i]][0]] + ((i - idxPre[nums[i]][0]) * idxPre[nums[i]][1])
                idxPre[nums[i]] = (i, idxPre[nums[i]][1] + 1)
            else:
                idxPre[nums[i]] = (i, 1)
        for i in range(len(nums) - 1, -1, -1):
            if nums[i] in idxSuf:
                suf[i] = suf[idxSuf[nums[i]][0]] + ((idxSuf[nums[i]][0] - i) * idxSuf[nums[i]][1])
                idxSuf[nums[i]] = (i, idxSuf[nums[i]][1] + 1)
            else:
                idxSuf[nums[i]] = (i, 1)
        for i  in range(len(nums)):
            res.append(pre[i] + suf[i])
        return res
        

File: 2615-sum-of-distances/2615-sum-of-distances.cpp
class Solution {
public:
    vector<long long> distance(vector<int>& nums) {
        vector<long long> res;
        unordered_map<int, pair<int, int>> idxPre;
        unordered_map<int, pair<int, int>> idxSuf;
        vector<long long> pre(nums.size(), 0);
        vector<long long> suf(nums.size(), 0);
        for (int i = 0; i < nums.size(); ++i) {
            if (idxPre.count(nums[i])) {
                pre[i] = pre[idxPre[nums[i]].first] + ((long long)(i - idxPre[nums[i]].first) * idxPre[nums[i]].second);
                idxPre[nums[i]] = {i, idxPre[nums[i]].second + 1};
            } else {
                idxPre[nums[i]] = {i, 1};
            }
        }
        for (int i = nums.size() - 1; i >= 0; --i) {
            if (idxSuf.count(nums[i])) {
                suf[i] = suf[idxSuf[nums[i]].first] + ((long long)(idxSuf[nums[i]].first - i) * idxSuf[nums[i]].second);
                idxSuf[nums[i]] = {i, idxSuf[nums[i]].second + 1};
            } else {
                idxSuf[nums[i]] = {i, 1};
            }
        }
        for (int i = 0; i < nums.size(); ++i) {
            res.push_back(pre[i] + suf[i]);
        }
        return res;
    }
};